File: Search for a Range.cpp

class Solution {
public:
    int findFirstGreater(int A[], int n, int target) {
        if (n == 0)
            return -1;
		if (n == 1)
			return (A[0] > target)?0:1;
        if (A[n-1] <= target)
            return n;
        if (A[0] > target)
            return 0;
        int start = 0;  // smaller or equal to target
        int end = n - 1;    // greater than target
        while (end - start > 1) {
            int mid = (start + end) / 2;
            if (A[mid] <= target)
                start = mid;
            else 
                end = mid;
        }
        return end;
    }
    vector<int> searchRange(int A[], int n, int target) {
        vector<int> ret(2);
        ret[0] = findFirstGreater(A, n, target - 1);
        ret[1] = findFirstGreater(A, n, target) - 1;
        if (ret[1] < ret[0])
            ret[0] = ret[1] = -1;
		else if (ret[0] == ret[1]) {
			if (ret[0] < 0 || A[ret[0]] != target)
				ret[0] = ret[1] = -1;
		}
        return ret;
    }
};

File: Search for a Range.java
public class Solution {
    public int findFirstGreater(int[] A, int target) {
        int upper = A.length;
        int lower = -1;
        while (upper - lower > 1) {
            int mid = (upper + lower) / 2;
            if (A[mid] > target)
                upper = mid;
            else
                lower = mid;
        }
        return upper;
    }
    public int[] searchRange(int[] A, int target) {
        int[] ret = new int [2];
        ret[0] = -1;
        ret[1] = -1;
        if (A.length == 0)
            return ret;
        
        int start = findFirstGreater(A, target - 1);
        if (start >= A.length)
            return ret;
            
        int end = findFirstGreater(A, target) - 1;
        if (end < start || end >= A.length)
            return ret;

        ret[0] = start;
        ret[1] = end;
        return ret;
    }
}